File: content/snippets/python/language_basics_functions_and_errors.py
#!/usr/bin/env python3
from typing import Optional


def add(a: int, b: int) -> int:
    return a + b


def safe_div(a: float, b: float) -> Optional[float]:
    if b == 0:
        return None
    return a / b


def strict_div(a: float, b: float) -> float:
    if b == 0:
        raise ValueError("division by zero")
    return a / b


def main() -> None:
    print("add(2, 3):", add(2, 3))
    print("safe_div(10, 2):", safe_div(10, 2))

    try:
        print("strict_div(10, 0):", strict_div(10, 0))
    except ValueError as err:
        print("error:", err)


if __name__ == "__main__":
    main()

File: content/snippets/typescript/language_basics_functions_and_errors.ts
#!/usr/bin/env ts-node

import { setImmediate } from "node:timers";

function add(a: number, b: number): number {
  return a + b;
}

function safeDiv(a: number, b: number): number | null {
  if (b === 0) return null;
  return a / b;
}

function strictDiv(a: number, b: number): number {
  if (b === 0) throw new Error("division by zero");
  return a / b;
}

function main(): void {
  console.log("add(2, 3):", add(2, 3));
  console.log("safeDiv(10, 2):", safeDiv(10, 2));

  try {
    console.log("strictDiv(10, 0):", strictDiv(10, 0));
  } catch (err) {
    console.error("error:", err);
  }
}

main();
setImmediate(() => {});
